File: c++/validParentheses#20/try1.cpp
// This is the optimal solution

class Solution {
public:
    bool isValid(string s) 
    {
        stack<char> theses;

        for(int i = 0; i < s.size(); i++)
        {
            if(s[i] == '(' || s[i] == '{' || s[i] == '[')
            {
                theses.push(s[i]);
            }
            else if((s[i] == ')' || s[i] == '}' || s[i] == ']') && theses.empty())
            {
                return false;
            }
            else if((s[i] == ')') && theses.top() != '(')
            {
                return false;
            }
            else if((s[i] == '}') && theses.top() != '{')
            {
                return false;
            }
            else if((s[i] == ']') && theses.top() != '[')
            {
                return false;
            }
            else
            {
                theses.pop();
            }
        }

        return theses.empty();
    }
};

// Time: O(n)
// Space: O(n)

File: python/subsetsII#90/subsetsII/try1.py
# this is the optimal solution

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()

        cur = []
        def backtracking(i):
            if i >= len(nums):
                res.append(copy.copy(cur))
                return

            cur.append(nums[i])
            backtracking(i + 1)
            while i + 1 < len(nums) and nums[i] == nums[i+1]:
                i += 1
            
            cur.pop()
            backtracking(i+1)

        backtracking(0)
        return res

# Time: O(n*n^2)
# Space: O(n)
